### PR TITLE
Gaussian Process Buckling Constraints in Blade Stiffened Shell Constitutive Subclass

### PR DESCRIPTION
**Pull request by sean-engelstad**
This code is linked to the repo, [ml_buckling](https://github.com/smdogroup/ml_buckling)

* Use Gaussian Processes (statistical machine learning approach) to learn buckling constraints
* Created a new subclass `TACSGPBladeStiffenedShellConstitutive` from the base class `TACSBladeStiffenedShellConstitutive`, which overrides the buckling constraints and combined buckling-stress failure criterion
* Buckling constraints include global + local modes for axial and shear (same as before except more accurate). Also stiffener crippling modes are treated as well.
* The `TACSBladeStiffenedShellConstitutive` used panel length as a DV which was enforced to be equal to the actual panel length from a separate computation of a panel length constraint. In the new class `TACSGPBladeStiffenedShellConstitutive`, panel width is also required, and both the panel length and width need to be set in at runtime through some additional infrastructure in TACS (not through a constraint => as this doesn't work great with our FUNtoFEM + ESP/CAPS infrastructure).

TODO:
- [x] Implement the machine learning buckling constraints (so far just have the closed-form ones implemented).
- [x] Update the failure criterion and sensitivity computations using the new buckling constraints.
- [x] Add panel width constraint in `tacs/constraints`
- [x] Add internal tests to the `TACSGPBladeStiffenedShellConstitutive` class for the following subtests:
    - [x] Nondimensional parameters
    - [x] Critical loads
    - [x] GP Model kernel and predictMeanTestData sens tests
- [x] Fix so all internal tests pass
    - [x] Closed-form case
    - [x] Machine learning case
    - [x] Add printLevel to each of these tests so you can turn it off
- [x] Fix so failure tests pass
    - [x] Add ply fraction derivs in DV failure test
    - [x] failStrain test
    - [x] failDV test 
- [x] Check the physical correctness of the smeared vs unsmeared stiffness computation when using the energy closed-form
     - [x] Double check if the `computePanelStiffness()` which uses smearedStiffnesses is correct or if we need a new unsmeared routine here.
     - [x] Add `computeStiffenerCripplingStiffness()` routine for panel-like stiffness of the stiffener 
- [x] Add cython objects for this class to be constructed.
- [x] Add full up test of the failure criterion and internal tests from the python level
- [x] Implement the panel length, width constraints as state variables for each TACS component in the FUNtoFEM TACS interface. Probably don't need a new state variable class, just store these values in lists representing each TACS component.
    - Compute a chain rule product of panel length, width variables to get the total shape derivatives for these state variables in FUNtoFEM and just update these state variables inside of FUNtoFEM. Don't want to expose these to the optimizer.
- [x] Add in the ML data and a way to read in this data using a classmethod in the cython probably.
- [x] Question : Should I change to using the smeared stiffener properties for the in plane load computation for global buckling? Answer: the actual in plane loads observed in the panel are what is computed in the local buckling section with B = 0 matrix. Smearing is done for global buckling to use an unstiffened panel closed-form approach. Of course I don't need to do that since I have a stiffened panel closed-form solution directly on the regular properties (and then later I use ML instead).
- [x] Additional tests for the constitutive class
    - [x] Add verification test of the buckling constraints (closed-form and ML) implemented in this class. This will make sure that they are giving the correct values (not just that the derivatives are correct).
    - [x] Time the constitutive class on the whole datasets and ensure the forward + adjoint eval time is reasonable. 
    - [x] If the time to compute Ytest on the ML models in each constitutive class don't scale well with 30k+ elements/constitutive objects => then I could register previous forward and adjoint inputs + outputs and return previous computed outputs.
    - [x]  Add shape derivative test over the panel length and width constraints
- [x] May also need to check whether the axial, shear, stiffener crippling data is duplicated for each element, constitutive object (I would not like to repeat and duplicate the data for each 30k elements if there are ~20k trials for ~100k floats per each trio of GP model objects as that would lead to about 3*10^9 floats of storage). There are ways to prevent an object from being copied, moved, etc. https://www.javatpoint.com/cpp-class-to-prevent-object-copies. [Answer: appears to not be duplicated as I use pointers according to @timryanb]
- [x] Add in sphinx and full documentation of each method
- [x] Rename `ShearGaussianProcessModel` and the other GP classes as `TACSShearGaussianProcessModel` etc.
- [x] Add a train routine feature so that if KS on the GPs are changed, it updates the weights by inverting a matrix (maybe in Cython / Python level)

### CODE REVIEW

**Comment by A-CGray** [comment]
Also @sean-engelstad , not sure how you're intending to do the panel length/width computation but you may be able to re-use what I have implemented in the [panel length constraint](https://github.com/smdogroup/tacs/blob/3c8bcdcd165fda9bb03012bfb48b379b2a1bfe6c/tacs/constraints/panel_length.py#L32-L76)

**Comment by sean-engelstad** [comment]
@A-CGray I decided to replicate what you did with the panel length constraint, with some small modifications for the panel width constraint. I will just add some code into FUNtoFEM to compute chain rule products of panel length, panel width DVs with the coordinate derivatives sensitivities for our workflow.

**Comment by sean-engelstad** [comment]
Update: internal tests for ML in the loop now pass! Next steps: test the failure criterion, etc. and other external tests on the constitutive object.
```
TACSGPBladeStiffened..testAllTests full results::
        testNondimensionalParameters = 1.8179e-07
        testAxialCriticalLoads = 1.3995e-08
        testShearCriticalLoads = 8.3006e-08
        testStiffenerCripplingLoad = 9.8267e-11
        testAxialGP all tests = 1.6610e-08
        testShearGP all tests = 1.6610e-08
        testCripplingGp all tests = 6.2022e-06
        Overall max rel error = 6.2022e-06
```

**Comment by sean-engelstad** [comment]
**Complex mode: failStrain test**
```
Testing the derivative of the failure index w.r.t. strain for object TACSBladeStiffenedShellConstitutive
Max Err: 2.1316e-14 in component 0.
Max REr: 1.0542e-13 in component 4.
          Val[   ]                  Analytic               Approximate                Rel. Error                Abs. Error
d(failure)/de[  0]   -3.8007334173280654e+01   -3.8007334173280633e+01    5.6084654544881153e-16    2.1316282072803006e-14
d(failure)/de[  1]   -3.2745208408550816e+01   -3.2745208408550795e+01    6.5097408472247376e-16    2.1316282072803006e-14
d(failure)/de[  2]    5.9636778818554667e+01    5.9636778818554667e+01    0.0000000000000000e+00    0.0000000000000000e+00
d(failure)/de[  3]    1.2952084768476078e+00    1.2952084768476078e+00    0.0000000000000000e+00    0.0000000000000000e+00
d(failure)/de[  4]   -1.0530991148876019e-03   -1.0530991148874908e-03    1.0542436214503594e-13    1.1102230246251565e-16
d(failure)/de[  5]   -1.0137600641731894e-03   -1.0137600641732471e-03    5.6896653966408030e-14    5.7679555576228836e-17
d(failure)/de[  6]    0.0000000000000000e+00    0.0000000000000000e+00                         -    0.0000000000000000e+00
d(failure)/de[  7]    0.0000000000000000e+00    0.0000000000000000e+00                         -    0.0000000000000000e+00
d(failure)/de[  8]    0.0000000000000000e+00    0.0000000000000000e+00                         -    0.0000000000000000e+00
```

**Comment by sean-engelstad** [comment]
All the failureDV tests pass now! A few small fixes with the strain transform computation from `panelStrain -> stiffenerStrain` did the trick. Another problem was a local variable `t` was supposed to be stiffener thick but was actually panelThick because wasn't redefined locally.

**All 10 tests pass now!** namely in the test script `test_gp_blade_shell_constitutive.py`

*Complex step test of failure DV test*
```
Max Err: 1.8865e-17 in component 2.
Max REr: 4.3805e-12 in component 3.
          Val[   ]                  Analytic               Approximate                Rel. Error                Abs. Error
d(failure)/dx[  0]   -3.2584854162225586e-08   -3.2584854162225474e-08    3.4524188062079851e-15    1.1249656330721176e-22
d(failure)/dx[  1]    8.0933749272162201e-06    8.0933749272153866e-06    1.0298304818370740e-13    8.3348042009823153e-19
d(failure)/dx[  2]   -1.9474018924833293e-03   -1.9474018924833482e-03    9.6873264188886932e-15    1.8865117801247777e-17
d(failure)/dx[  3]    6.7302764114468927e-07    6.7302764114174107e-07    4.3805096132984838e-12    2.9482040520369991e-18
d(failure)/dx[  4]    1.1597892968325800e-05    1.1597892968325456e-05    2.9651539079076256e-14    3.4389537658524594e-19
d(failure)/dx[  5]    1.6880613492434901e-06    1.6880613492433107e-06    1.0625160433979342e-13    1.7935922658109810e-19
d(failure)/dx[  6]   -4.5232780788233059e-03   -4.5232780788233033e-03    5.7526536476884563e-16    2.6020852139652106e-18
d(failure)/dx[  7]   -4.5084343173967456e-03   -4.5084343173967430e-03    5.7715939299027094e-16    2.6020852139652106e-18
d(failure)/dx[  8]   -3.4248925127256912e-06   -3.4248925127247099e-06    2.8651634051237215e-13    9.8128766939410694e-19
d(failure)/dx[  9]   -6.8563572486820331e-07   -6.8563572486809235e-07    1.6183712730497074e-13    1.1096131609031334e-19
d(failure)/dx[ 10]    6.4420866224409167e-06    6.4420866224412454e-06    5.1015890160465777e-14    3.2864878353466853e-19
```

**Comment by sean-engelstad** [comment]
Fixed a problem with the Newton iteration abs tolerance (too strict and couldn't converge due to floating point limit). Switched to relative tolerance and added max_iter limit for Newton iteration in closed-form solution for shear critical loads.

**Comment by sean-engelstad** [comment]
@A-CGray @timryanb I merged in your code Ali and tried just overwriting the `evalLocalPanelBuckling`, `evalGlobalPanelBuckling` methods to use the base class `computeFailureValues` method. However, this method is not virtual so even though I define new `evalLocalPanelBuckling` methods, it does not call my buckling load predictions unless I copy and paste the same `computeFailureValues` code into my class as well (but then we've duplicated the code). 

Maybe you can store `std::function` pointers to the functions used for the buckling load predictions => and use those in the `computeFailureValues` method in the base class. And then I can change the value in the function pointers in my subclass

**Comment by sean-engelstad** [comment]
I recovered the same buckling load predictions as before (for low aspect ratios) with the new implementation by overwriting Ali's methods (also I did use the functors and lambda expressions approach).
```
N1CritGLobal = 6.05762950e+06
N12CritGlobal = 1.85230893e+07
```

**Comment by sean-engelstad** [comment]
I'm ready for you guys to review it now @timryanb @A-CGray

**Review by timryanb — CHANGES_REQUESTED** [review]
Can you make sure that all classes/methods added to constitutive.pyx have proper docstrings, describing what they do, inputs, outputs etc. See the other classes/methods for formatting reference

**Comment by sean-engelstad** [comment]
@timryanb I finished adding the docstrings for all methods, classes in the constitutive.pyx file that I added.

**Review by timryanb — APPROVED** [review]

**Comment by timryanb** [comment]
@A-CGray, did you have a chance to take a look at this?

File: tacs/constraints/panel_width.py
Comment (A-CGray) [inline tacs/constraints/panel_width.py:150]: These are the only lines that actually need to be different in the panel width constraint compared to the panel length constraint. I think we can reduce the amount of code in this class massively by putting this in a function called something like `computeRefAxis` just having the panel width constraint override that. Then there'd be no need to override `evalConstraints` and `evalConstraintSens` either.

I'm imagining something like this:

```python
refAxes.append(self.computeRefAxis(boundaryNodeCoords[compID], refAxis)
```

**Review by A-CGray — CHANGES_REQUESTED** [review]
Hey @sean-engelstad this is really good work. I have a lot of comments but they're mostly duplicates about using `new` and `delete[]` in places that you don't need to. I commented all the examples of this that I saw in the constitutive class, I think there are some more cases in the GPModel classes that I didn't leave comments on but should still be fixed.

Also, you could add a panel width constraint to [this integration test](https://github.com/smdogroup/tacs/blob/master/tests/integration_tests/test_shell_blade_stiffened_plate_quad.py) that already tests the panel length constraint

File: src/constitutive/TACSBladeStiffenedShellConstitutive.h
Comment (sean-engelstad) [reply src/constitutive/TACSBladeStiffenedShellConstitutive.h:21]: Yes good catch, removed it

File: src/constitutive/TACSGPBladeStiffenedShellConstitutive.h
Comment (sean-engelstad) [reply src/constitutive/TACSGPBladeStiffenedShellConstitutive.h:109]: Ok I updated this and removed any redundancies from your class here

File: src/constitutive/TACSGPBladeStiffenedShellConstitutive.h
Comment (sean-engelstad) [reply src/constitutive/TACSGPBladeStiffenedShellConstitutive.h:1070]: Yes it is a fixed constant, updated this

**Comment by sean-engelstad** [comment]
> Hey @sean-engelstad this is really good work. I have a lot of comments but they're mostly duplicates about using `new` and `delete[]` in places that you don't need to. I commented all the examples of this that I saw in the constitutive class, I think there are some more cases in the GPModel classes that I didn't leave comments on but should still be fixed.
> 
> Also, you could add a panel width constraint to [this integration test](https://github.com/smdogroup/tacs/blob/master/tests/integration_tests/test_shell_blade_stiffened_plate_quad.py) that already tests the panel length constraint

Thanks Ali, I'm resolving the comment discussions as I go through and update the code if that's good with you. For comments where there might be more discussion I'll leave it unresolved

File: src/constitutive/TACSGPBladeStiffenedShellConstitutive.h
Comment (sean-engelstad) [reply src/constitutive/TACSGPBladeStiffenedShellConstitutive.h:82]: Yeah I don't know why it says that lol, definitely still a design variable

File: src/constitutive/TACSGPBladeStiffenedShellConstitutive.h
Comment (sean-engelstad) [reply src/constitutive/TACSGPBladeStiffenedShellConstitutive.h:246]: I did want to use this occasionally to view non-dimensional parameters like `rho_0`, `gamma`, `xi`, when I was debugging using the ML model in the full wing. Leaving this comment open for discussion @A-CGray

File: src/constitutive/TACSGPBladeStiffenedShellConstitutive.h
Comment (sean-engelstad) [reply src/constitutive/TACSGPBladeStiffenedShellConstitutive.h:493]: Added these to the test routines

File: src/constitutive/TACSGPBladeStiffenedShellConstitutive.h
Comment (sean-engelstad) [reply src/constitutive/TACSGPBladeStiffenedShellConstitutive.h:721]: @A-CGray, I added docstrings to each of the header files I've added now, so resolving those conversations

File: src/constitutive/TACSGPBladeStiffenedShellConstitutive.cpp
Comment (sean-engelstad) [reply src/constitutive/TACSGPBladeStiffenedShellConstitutive.cpp:475]: Good point, fixed this one

File: src/constitutive/TACSGPBladeStiffenedShellConstitutive.cpp
Comment (sean-engelstad) [reply src/constitutive/TACSGPBladeStiffenedShellConstitutive.cpp:1555]: Ok thanks, this is good advice. I went ahead and changed this all throughout the TACSGPBladeStiffenedShellConstitutive class

File: src/constitutive/TACSGPBladeStiffenedShellConstitutive.cpp
Comment (sean-engelstad) [reply src/constitutive/TACSGPBladeStiffenedShellConstitutive.cpp:3994]: Sure

File: src/constitutive/TACSGaussianProcessModel.h
Comment (sean-engelstad) [reply src/constitutive/TACSGaussianProcessModel.h:53]: It's used in the kernel functions for soft_relu

File: src/constitutive/TACSGaussianProcessModel.h
Comment (sean-engelstad) [reply src/constitutive/TACSGaussianProcessModel.h:104]: Good idea

File: src/constitutive/TACSGaussianProcessModel.h
Comment (sean-engelstad) [reply src/constitutive/TACSGaussianProcessModel.h:162]: That's true, I set it up this way because I was originally testing out different kernel functions for each of them : AxialGP, ShearGP, CripplingGP; especially the CripplingGP. Then it ended up that I just stuck with the same kernel for both the AxialGP and ShearGP and the CripplingGP was less important. I guess I will change it then to make it cleaner

File: tests/constitutive_tests/test_gp_blade_shell_constitutive.py
Comment (sean-engelstad) [reply tests/constitutive_tests/test_gp_blade_shell_constitutive.py:341]: Ok I fixed this as well

File: tacs/constraints/panel_width.py
Comment (sean-engelstad) [reply tacs/constraints/panel_width.py:150]: Yeah good point, just cleaned it up

File: examples/gp_stiffened_plate/eval_stiffened_plate.py
Comment (sean-engelstad) [reply examples/gp_stiffened_plate/eval_stiffened_plate.py:1]: Combined it and just removed my old example files (don't want to add lots of extra example folders to TACS if not needed)

File: tacs/constitutive.pyx
Comment (sean-engelstad) [reply tacs/constitutive.pyx:1347]: fixed this

File: tacs/constitutive.pyx
Comment (sean-engelstad) [reply tacs/constitutive.pyx:1875]: @timryanb, @A-CGray It gave me a bunch of cython errors when I tried to inherit from the BladeStiffenedShellConstitutive class and do the constructor. I don't remember exactly why though. Definitely would be cleaner to not repeat these methods that should inherit from the super class

**Comment by sean-engelstad** [comment]
@A-CGray @timryanb, alright I'm done addressing all the items in Ali's review. The only point left up for debate is whether I should change my constitutive object to be a subclass of Ali's in constitutive.pyx which would cleanup a few repeated methods. It caused me quite a bit of problems last time I tried that, but maybe I would be able to figure it out now since I have more experience with the Cython.

File: src/constitutive/TACSGaussianProcessModel.h
Comment (A-CGray) [reply src/constitutive/TACSGaussianProcessModel.h:106]: I meant the `TacsScalar one = 1.0;` line specifically not the function in general

**Review by A-CGray — APPROVED** [review]
Thanks @sean-engelstad , this addresses all my comments, with the exception of the cython wrapper like you said, so I'd like @timryanb to give final approval on that.

I did a little reading of [the cython documentation](https://cython.readthedocs.io/en/latest/src/userguide/special_methods.html#initialisation-methods-cinit-and-init) and it seems like you can't override the __cinit__ of a parent class, so it will always be called, which causes issues here because we don't want to actually create a `BladeStiffenedShell` object when trying to create a `GPBladeStiffenedShell` object. 

@timryanb would it work if we switched these classes from using `__cinit__` to `__init__`?

**Comment by timryanb** [comment]
@sean-engelstad @A-CGray, I believe it should be possible to make this class into a cython sub-class. Both the GP and regular version of the cython class would have to inherit from a parent base class that held the common API between the two classes.

An example of this would be the `ShellConstitutive` parent class that all shell constitutive classes inherit from (`IsoShellConstitutive`, `CompShellConstitutive`, etc.).

In this case you could add the new common baseclass to the `constitutive.pxd` file like below:
```
cdef class StiffenedShellConstitutive(ShellConstitutive):
    cdef TACSBladeStiffenedShellConstitutive *base_ptr
    
 cdef class BladeStiffenedShellConstitutive(StiffenedShellConstitutive):
    cdef TACSBladeStiffenedShellConstitutive *blade_ptr
    
cdef class GPBladeStiffenedShell(StiffenedShellConstitutive):
    cdef TACS GPBladeStiffenedShell *blade_ptr
```

You can then add the API to the `constitutive.pyx` file using `ShellConstitutive` as a template (see [here](https://github.com/smdogroup/tacs/blob/master/tacs/constitutive.pyx#L677-L691))

**Comment by sean-engelstad** [comment]
Ok @timryanb @A-CGray  => I removed the redundant code in constitutive.pyx using Tim's suggestions. As long as these tests pass, it should be good to go now.

**Comment by sean-engelstad** [comment]
Wait => don't merge it yet, there is a problem when running on big cases right now which I've traced back to incref()

**Comment by sean-engelstad** [comment]
Update: I needed to also store one of the other constitutive subclass ptrs namely `cptr` otherwise constructing any elements (from element.pyx) will fail in `self.cptr.incref()`. So I added the copy `self.cptr = ` back in the constructors

**Comment by sean-engelstad** [comment]
@timryanb should be good to merge now if you wanna take one more look. I can run the wing examples and the tests pass now with the updates to constitutive.pyx

**Comment by timryanb** [comment]
I should be able to take a look at it sometime tomorrow @sean-engelstad

**Comment by timryanb** [comment]
@sean-engelstad also can you add your new panel constraints to the constraint autodocs list [here](https://github.com/smdogroup/tacs/blob/master/docs/source/pytacs/constraints.rst). You just have to add the name of the python file in the constraints module folder that your new code lives in

**Comment by sean-engelstad** [comment]
@timryanb - ok I added the docs and fixed constitutive.pyx

File: docs/source/pytacs/panel_width.rst
Comment (timryanb) [inline docs/source/pytacs/panel_width.rst:1]: Just change this title to PanelWidthConstraint and we're good

**Comment by sean-engelstad** [comment]
@timryanb I fixed the docs mistake you just caught

**Pull request merged**
